cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/type/DirectoryClassLoaderProvider.java

```diff
@@ -0,0 +1,106 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.type;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.lang.jar.BundleJarUtil;
+import co.cask.cdap.common.utils.DirUtils;
+import co.cask.cdap.proto.DatasetModuleMeta;
+import co.cask.common.lang.DirectoryClassLoader;
+import com.google.common.base.Objects;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import javax.annotation.Nullable;
+
+/**
+ * Creates a {@link ClassLoader} for a {@link DatasetModuleMeta} by unpacking the dataset jar and creating a
+ * {@link DirectoryClassLoader} over the unpacked jar. Classloaders are cached, and unpacked directories are cleaned
+ * up when the provider is closed. Note that this means changes to dataset code are not picked up, as the assumption
+ * is that this provider is created once at the start of a program run and closed at the end.
+ */
+public class DirectoryClassLoaderProvider implements DatasetClassLoaderProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(DirectoryClassLoaderProvider.class);
+  private final ClassLoader parentClassLoader;
+  private final LoadingCache<URI, ClassLoader> classLoaders;
+  private final LocationFactory locationFactory;
+  private final File tmpDir;
+
+  public DirectoryClassLoaderProvider(CConfiguration cConf,
+                                      @Nullable ClassLoader parentClassLoader,
+                                      LocationFactory locationFactory) {
+    this.parentClassLoader = parentClassLoader == null ?
+      Objects.firstNonNull(Thread.currentThread().getContextClassLoader(), getClass().getClassLoader()) :
+      parentClassLoader;
+    this.locationFactory = locationFactory;
+    this.classLoaders = CacheBuilder.newBuilder().build(new ClassLoaderCacheLoader());
+    File baseDir =
+      new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR), cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    this.tmpDir = DirUtils.createTempDir(baseDir);
+  }
+
+  @Override
+  public ClassLoader get(DatasetModuleMeta moduleMeta) throws IOException {
+    URI jarLocation = moduleMeta.getJarLocation();
+    return jarLocation == null ? parentClassLoader : classLoaders.getUnchecked(jarLocation);
+  }
+
+  @Override
+  public ClassLoader getParent() {
+    return parentClassLoader;
+  }
+
+  @Override
+  public void close() throws IOException {
+    // Cleanup the ClassLoader cache and the directory used for expanded dataset jars
+    classLoaders.invalidateAll();
```

Comment:
Need to have a removal listener to close the ClassLoaders as well (if the ClassLoader is closable, which usually is). See https://github.com/caskdata/cdap/blob/develop/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/PluginInstantiator.java#L82